sklearn/cluster/_kmeans.py

```diff
@@ -256,15 +257,148 @@ def _kmeans_plusplus(X, n_clusters, x_squared_norms, random_state, n_local_trial
+    """
+
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def prepare_fit(self, X, y=None, sample_weight=None):
+        estimator = self.estimator
+        X = estimator._validate_data(
+            X,
+            accept_sparse="csr",
+            dtype=[np.float64, np.float32],
+            order="C",
+            copy=estimator.copy_x,
+            accept_large_sparse=False,
+        )
+        # this sets estimator _algorithm implicitly
+        # XXX: shall we explose this logic as part of then engine API?
+        # or is the current API flexible enough?
+        estimator._check_params_vs_input(X)
+
+        # TODO: delegate rng and sample weight checks to engine
+        random_state = check_random_state(estimator.random_state)
+        sample_weight = _check_sample_weight(sample_weight, X, dtype=X.dtype)
+
+        # Also store the number of threads on the estimator to be reused at
+        # prediction time XXX: shall we wrap engine-specific private fit
+        # attributes in a predict context dict set as attribute on the
+        # estimator?
+        estimator._n_threads = self._n_threads = _openmp_effective_n_threads()
+
+        # Validate init array
+        init = estimator.init
+        init_is_array_like = _is_arraylike_not_scalar(init)
+        if init_is_array_like:
+            init = check_array(init, dtype=X.dtype, copy=True, order="C")
+            estimator._validate_center_shape(X, init)
+
+        # subtract of mean of x for more accurate distance computations
+        if not sp.issparse(X):
+            X_mean = X.mean(axis=0)
+            # The copy was already done above
+            X -= X_mean
+
+            if init_is_array_like:
+                init -= X_mean
+
+            self.X_mean = X_mean
+
+        # precompute squared norms of data points
+        x_squared_norms = row_norms(X, squared=True)
+
+        if estimator._algorithm == "elkan":
+            kmeans_single = _kmeans_single_elkan
+        else:
+            kmeans_single = _kmeans_single_lloyd
+            estimator._check_mkl_vcomp(X, X.shape[0])
+
+        self.x_squared_norms = x_squared_norms
+        self.kmeans_single_func = kmeans_single
+        self.random_state = random_state
+        self.tol = self.scale_tolerance(X, estimator.tol)
+        self.init = init
+        return X, y, sample_weight
+
+    def init_centroids(self, X):
+        # XXX: the actual implementation of the centroids init should also be
+        # moved to the engine.
+        return self.estimator._init_centroids(
+            X,
+            x_squared_norms=self.x_squared_norms,
+            init=self.init,
+            random_state=self.random_state,
+        )
+
+    def scale_tolerance(self, X, tol):
+        """Return a tolerance which is dependent on the dataset."""
+        if tol == 0:
+            return 0
+        if sp.issparse(X):
+            variances = mean_variance_axis(X, axis=0)[1]
```

Comment:
I find this is a bit more understandable: it makes explicit that `means` and `sum_weights` are ignored. 
```suggestion
            _, variances, _ = mean_variance_axis(X, axis=0)
```

What do you think?